event-store-plans.md

```diff
@@ -0,0 +1,174 @@
+low hanging fruit because all you'd do is switch up the cached `IDocumentStorage` objects for a certain document type.
+
+
+## Ensure that aggregates updated "Inline" are coming from identity map
+
+Imagine this very common scenario:
+
+A CQRS command handler needs to:
+
+1. Fetch an aggregate of an event stream as a "write model" to process the command
+2. Emit some events
+3. Maybe update the projected aggregation if it's running `Inline`
+4. For whatever reason, access the current state of the aggregate again in the handler. Maybe because it's a response type in an HTTP handler, maybe to transmit
+   messages, who knows
+
+Maybe you even want the updated aggregate to make some kind of sense of the next part of the work -- which I'd at least push Wolverine folks to do with cascading messages, but still
+
+Regardless of the Marten `IDocumentSession` type, I think it would be valuable to **prevent** unnecessary extra reads of the aggregate, especially if there's some
+re-application of the events. 
+
+At a minimum, I think that the aggregation fetching should use the identity map *even if the session itself isn't using the identity map*. See the next item.
+This probably impacts both the `FetchForWriting()` behavior and the inline projection behavior
+
+This absolutely impacts the Wolverine "aggregate handler workflow". There's an opportunity to eliminate some serious inefficiency today
+
+
+## Versioned Projections for Blue/Green deployment
+
+This is really just sticking a "projection version number" on the Projection subclasses that would potentially be used as a suffix on the document
+tables for the projected documents (the document alias). This would enable blue/green deployment of the same projection so that nodes on the previous version can be 
+using the previous revision/definition of the projection while other nodes are running the newer model. 
+
+Obviously there's some serious concern about just how the new projection gets rebuilt in time. I think the "catch up" mode that this document
+introduces as an "Async strategy for FetchForWriting()" is necessary to make this viable
+
+
+## "Poor Man's Load Balancing" of the Async Daemon
+
+See Oskar's post [How to scale out Marten](https://event-driven.io/en/scaling_out_marten/). I'm voting to just bake this work directly into Marten as a 
+first class thing. Do some randomness of the hot/cold behavior for individual projection groups to see if other nodes can take up that work.
+
+I think the async daemon for each tenant database in the case of multi-tenancy through separate databases should be allowed to run on separate nodes. Might take some 
+work with randomness to kind of distribute the load per tenant
+
+By no means will this be a true equivalent of the projection load balancing work planned for "CritterStackPro"
+
+## Dynamic Database per Tenant Multi-Tenancy model
```

Comment:
This is great to see, and keen to try it though I dont know how to deal with the per-tenant db scaling issues.

Context: 
In one aws region we have ~300 tenants. At this time our app uses conjoined tenancy on an "r6g.xl" instance size. To split them up would either mean creating ~300 rds instances. By default the instance limit is 40 per region, and even ignoring that the cost per instance would skyrocket as they would be forced to be massively over-provisioned instances (smallest viable size is "small", due to below issue).

The alternative is to group tenants together onto a smaller set of instances, say 50 per rds instance.  This introduces a new problem where the memory overhead of connections starts to add up and consume all memory in the database. The memory consumed (by default) would be the mem allocated per connection (8mb) * max connections by app (100) * tenants in server (50) * nodes in app (2-4), resulting in >80gb memory used in this situation. This puts scaling limits on your application and forces the db to be over-provisioned.

A possible workaround is Amazon RDS Proxy. Previously this wasn't a good solution as the postgres support was always several major versions behind current, but now it seems to be in sync with postgres 16.  There are a [bunch of other restrictions](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/rds-proxy.html#rds-proxy.limitations-pg) that will necessitate testing though.


This feels like a "postgres" or "aws" problem rather than a "marten" problem, but it would be interesting to see how other clients tackle it given this work is sponsored by clients.

Reply:
I think that one option is to also add an option per schema or mention in other places partitioning per tenant or some other way to allow other shared contexts (or mixed option).

I think that a database per tenant is useful when:
- we have a limited (and short) number of tenants,
- the legal requirement to prove data isolation,
- ops team to maintain it.

When databases number grows, as you described it gets extremely hard and costly from the DevOps perspective to maintain that. I wouldn't consider a database per tenant as a viable solution for a big SaaS platform.